src/Topologies/Topologies.jl

```diff
@@ -166,10 +166,318 @@ struct Vertex{T <: AbstractTopology, V}
     num::V
 end
 
+"""
+    TensorProductTopology(mesh)
+
+A generic tensor-product topology defined on either an equispaced, regular rectangular
+structured grid or irregular one.
+"""
+struct TensorProductTopology{M <: AbstractMesh, B} <: AbstractTopology
+    mesh::M
+    boundaries::B
+end
+function TensorProductTopology(mesh::M) where {M <: AbstractMesh}
+    x1boundary = mesh.domain.x1boundary
+    x2boundary = mesh.domain.x2boundary
+    boundaries = if isnothing(x1boundary)
+        if isnothing(x2boundary)
+            NamedTuple()
+        else
+            NamedTuple{x2boundary}((3, 4))
+        end
+    else
+        if isnothing(x2boundary)
+            NamedTuple{x1boundary}((1, 2))
+        else
+            NamedTuple{(x1boundary..., x2boundary...)}((1, 2, 3, 4))
+        end
+    end
+    TensorProductTopology(mesh, boundaries)
+end
+
+
+
+function Base.show(
+    io::IO,
+    topology::TensorProductTopology{M},
+) where {M <: AbstractMesh}
+    print(io, "TensorProductTopology on ", topology.mesh)
+end
+domain(topology::TensorProductTopology{M}) where {M <: AbstractMesh} =
+    topology.mesh.domain
+
+function nlocalelems(
+    topology::TensorProductTopology{M},
+) where {M <: AbstractMesh}
+    n1 = topology.mesh.n1
+    n2 = topology.mesh.n2
+    return n1 * n2
+end
+
+eachslabindex(topology::TensorProductTopology{M}) where {M <: AbstractMesh} =
+    1:nlocalelems(topology)
+
+
+# InteriorFaceIterator
+function Base.length(fiter::InteriorFaceIterator)
```

Comment:
As per @charleskawczynski 's suggestion on Slack, this and the following methods for the iterators could work as well with the following signature:
```
Base.length(fiter::InteriorFaceIterator{T}) where {M <: AbstractMesh, T <: TensorProductTopology{M}}
```
@simonbyrne , Since both Charlie's way and the one that I used here (i.e., without any type specification) work and pass all tests, is it advisable to change it for a better type safety, or it is really not that important and can be left as is now?

Reply:
I decided to go the safest way in [a1607c7](https://github.com/CliMA/ClimaCore.jl/pull/82/commits/a1607c7e26483e592f57847facbd8288e7fb6ff1) and followed @charleskawczynski 's suggested way :)